### ISSUE TITLE
Failure on inlines with zero existing rows

### ISSUE THREAD

**Issue by murraylchapman** [crucialfelix/django-ajax-selects#170]
I'm seeing a problem when using ajax_select in a TabularInline form that has no existing rows.

The problem is that ajax_select attaches the jQuery-UI autocomplete widget to the input field in the hidden row and when that gets cloned to add a row the internal variables that jQuery uses for housekeeping aren't updated to point to the cloned row.  I've encountered a similar problem when using autoNumeric; the only way to get it to work is make sure you don't attach stuff to anything in the hidden row; you've got to do it _after_ cloning.

I _think_ that's what the "dirty loop" is designed to compensate for; it seems to work somehow when there's existing rows, but it's hard to tell because I'm not sure what it's actually attempting to do.  Certainly when there's zero existing rows the "minus 2" logic results in an html_id that doesn't map to anything.

I was able to fix things by replacing the "dirty loop" with this:

```
/* Don't attach to fields in the hidden row; we attach upon row add  */
if (html_id.indexOf('__prefix__') !== -1) {
    return;
}
```

...which seems to work in all cases I tried it in.  I'm not sure if there are unintended side-effects because I'm not really sure what the loop is trying to do.

(Django 1.7, python 2.7)

**Comment by crucialfelix** [comment]
That particular part was submitted by somebody. It is not clean. I'm not a fan of jQuery or jQuery ui either.

Is there one of the forms in the example app that resembles your test case ? TabularInline form, foreign key ? m2m ?

**Comment by crucialfelix** [comment]
> ...which seems to work in all cases I tried it in. I'm not sure if there are unintended side-effects because I'm not really sure what the loop is trying to do.

which is how we got that messy piece of code in there to begin with ;)

**Comment by murraylchapman** [comment]
I downloaded the sample app via the instructions in the docs and ran it, and couldn't reproduce the bug, which was enough for me to find the source of the problem.

The problem is in the interaction with Firefox and Django's use of hidden form fields.  Firefox likes to cache form fields upon reload; the "autcomplete=off" on the field is supposed to prevent this but at least for my setup it's not.  This is a Firefox specific issue.  Further, Django changed in 1.8.8+ to send stronger cache control headers ("no-store") that stops Firefox behaving in this way.  There's a field (`id_foo_set-TOTAL_FORMS`) that javascript updates when you add rows, and when you hit reload in Firefox the updated value sticks around rather than being reset.  That's bad.

Aside: I tried reinstalling the example app with Django 1.7 to confirm this, but it always seems to be running with 1.9.  I don't know enough about virtualenv to figure out why.

But this "dirty loop" seems to be associated with the problem; I'd like to try and figure out what it thinks it's doing:

    /* detects inline forms and converts the html_id if needed */
    if (html_id.indexOf('__prefix__') !== -1) {
      // Some dirty loop to find the appropriate element to apply the callback to
      while ($('#' + html_id).length) {
        html_id = prefix_id.replace(/__prefix__/, prefix++);
      }
      html_id = prefix_id.replace(/__prefix__/, prefix - 2);
      // Ignore the first call to this function, the one that is triggered when
      // page is loaded just because the 'empty' form is there.
      if ($('#' + html_id + ', #' + html_id + '_text').hasClass('ui-autocomplete-input')) {
        return;
      }
    }

This code is run on every input in the form, including the ones in the hidden row.  The top-level `if` statement detects when we're looking at a field in the hidden row (the presence of `__prefix__`) and then does *something* to decide whether to optionally `return` to avoid attaching autocomplete (again?).  The only thing that this code can do is cause the routine to bail out early.

Looking at the commit history, this code was added before the code that hooks and triggers upon the creation of new rows, so I think it's actually redundant now.  I *think* it was an attempt to do something similar via jiggery-pokery on the html_ids.

So in summary, I think you can just eliminate this whole "dirty loop" as I suggested above and just `return` when you're looking at the hidden row; IMHO you've got a much better and more elegant solution for dynamically added rows by hooking the add-row functionality.

**Comment by crucialfelix** [comment]
Thanks for the in depth research ! Nicely done.

If I had time I would set up selenium and then I could verify this complex interaction (jqueryui, several djangos, several browsers, the admin, these little hacks).

Really I don't even use the django-admin if I can avoid it because it is a rats nest.

**Comment by cheluis** [comment]
@murraylchapman You can install specific Django versions with pip, try uninstalling your current django version with `pip uninstall Django` and then `pip install Django==1.9`

**Comment by crucialfelix** [comment]
The install.sh script in examples does this for you 

    ./install.sh 1.9

**Comment by crucialfelix** [comment]
I understand that you are seeing a problem of some sort, but I still don't know what that problem actually manifests as and what I can do to see it.

I'm using Firefox and Django 1.6 .. 1.10 and I'm adding rows, filling out things, editing things and it all works fine. I'm adding and editing Authors http://localhost:8000/admin/example/author/

I also have tested that the "dirty-loop" doesn't do anything useful anymore and should probably never have been in there.

However I don't see any problems with having activated the __prefix__ inputs (the ones that are cloned when you "add another Book"). Those are just handlers for events, they don't cache anything.

I need to know what you are experiencing so I can verify that it is a problem and that it is caused by activating the __prefix__ inputs.

**Comment by crucialfelix** [comment]
In the 1.5.0 release: 
* I have removed the dirty loop. 
* It does not activate `__prefix__` fields. 
* It does not double activate fields. 

The last was causing multiple handlers to be added to fields whenever you click "add another..." and this was causing multiple objects to be added to manytomany inline fields. (because there were multiple handlers a-handlin')

This may have solved whatever issues you were seeing.

Please let me know (when you get a chance) exactly what you were seeing.

**Comment by murraylchapman** [comment]
Yes, this is now fixed.  Thanks.
